src/components/intent.ts

```diff
@@ -0,0 +1,829 @@
+        else {
+            this._membershipStates = {
+            //  room_id : "join|invite|leave|null"   null=unknown
+            };
+            this._powerLevels = {
+            //  room_id: event.content
+            };
+        }
+        this.opts = {
+            ...opts,
+            backingStore: opts.backingStore || {
+                getMembership: (roomId: string, userId: string) => {
+                    if (userId !== this.client.credentials.userId) {
+                        return null;
+                    }
+                    return this._membershipStates![roomId];
+                },
+                getPowerLevelContent: (roomId: string) => {
+                    return this._powerLevels![roomId];
+                },
+                setMembership: (roomId: string, userId: string, membership: MembershipState) => {
+                    if (userId !== this.client.credentials.userId) {
+                        return;
+                    }
+                    this._membershipStates![roomId] = membership;
+                },
+                setPowerLevelContent: (roomId: string, content: Record<string, unknown>) => {
+                    this._powerLevels![roomId] = content;
+                },
+            },
+            caching: {
+                size: opts.caching?.ttl || DEFAULT_CACHE_SIZE,
+                ttl: opts.caching?.ttl || DEFAULT_CACHE_TTL,
+            },
+            enablePresence: opts.enablePresence !== false,
+        };
+        this._requestCaches = {
+            profile: new ClientRequestCache(
+                this.opts.caching.ttl,
+                this.opts.caching.size,
+                (_: any, userId: string, info: string) => {
+                    return this.getProfileInfo(userId, info, false);
+                }
+            ),
+            roomstate: new ClientRequestCache(
+                this.opts.caching.ttl,
+                this.opts.caching.size,
+                (roomId: string) => {
+                    return this.roomState(roomId, false);
+                }
+            ),
+            event: new ClientRequestCache(
+                this.opts.caching.ttl,
+                this.opts.caching.size,
+                (_: any, roomId: string, eventId: string) => {
+                    return this.getEvent(roomId, eventId, false);
+                }
+            ),
+        };
+    }
+
+    /**
+     * Return the client this Intent is acting on behalf of.
+     * @return The client
+     */
+    getClient = () => {
+        return this.client;
+    };
+
+    /**
+     * <p>Send a plaintext message to a room.</p>
+     * This will automatically make the client join the room so they can send the
+     * message if they are not already joined. It will also make sure that the client
+     * has sufficient power level to do this.
+     * @param roomId The room to send to.
+     * @param text The text string to send.
+     */
+    sendText = (roomId: string, text: string) => {
```

Comment:
Could these be `public sendText` methods?

Reply:
Are they not public by default?

Reply:
Yes but it looks very ugly to use the  `func = (args)`. Can we drop the `=`. I'm leaning towards including `public` for readability, but I'm not so bothered about it.

Reply:
Yeah, we can drop `func = (args) => {}`. I usually use arrow functions whenever I can, because I don't trust `this` in `function() {}` anymore.

Reply:
You can trust class methods :). They are here for you.